FreeAPS/Sources/Services/ContactTrick/ContactTrickManager.swift

```diff
@@ -0,0 +1,196 @@
+import Combine
+import Contacts
+import Foundation
+import Swinject
+
+protocol ContactTrickManager {
+    func updateContacts(contacts: [ContactTrickEntry], completion: @escaping (Result<Void, Error>) -> Void)
+}
+
+final class BaseContactTrickManager: NSObject, ContactTrickManager, Injectable {
+    private let processQueue = DispatchQueue(label: "BaseContactTrickManager.processQueue")
+    private var state = ContactTrickState()
+    private let contactStore = CNContactStore()
+    private var workItem: DispatchWorkItem?
+
+    @Injected() private var broadcaster: Broadcaster!
+    @Injected() private var settingsManager: SettingsManager!
+    @Injected() private var apsManager: APSManager!
+    @Injected() private var storage: FileStorage!
+
+    private var contacts: [ContactTrickEntry] = []
+
+    let coreDataStorage = CoreDataStorage()
+
+    init(resolver: Resolver) {
+        super.init()
+        injectServices(resolver)
+
+        contacts = storage.retrieve(OpenAPS.Settings.contactTrick, as: [ContactTrickEntry].self)
+            ?? [ContactTrickEntry](from: OpenAPS.defaults(for: OpenAPS.Settings.contactTrick))
+            ?? []
+
+        broadcaster.register(GlucoseObserver.self, observer: self)
+        broadcaster.register(SuggestionObserver.self, observer: self)
+        broadcaster.register(SettingsObserver.self, observer: self)
+
+        configureState()
+    }
+
+    func updateContacts(contacts: [ContactTrickEntry], completion: @escaping (Result<Void, Error>) -> Void) {
+        processQueue.async {
+            self.contacts = contacts
+            self.renderContacts()
+            completion(.success(()))
+        }
+    }
+
+    private func configureState() {
+        processQueue.async {
+            let readings = self.coreDataStorage.fetchGlucose(interval: DateFilter().twoHours)
+            let glucoseValues = self.glucoseText(readings)
+            self.state.glucose = glucoseValues.glucose
+            self.state.trend = glucoseValues.trend
+            self.state.delta = glucoseValues.delta
+            self.state.glucoseDate = readings.first?.date ?? .distantPast
+            self.state.lastLoopDate = self.apsManager.lastLoopDate
+
+            self.state.iob = self.suggestion?.iob
+            self.state.cob = self.suggestion?.cob
+            self.state.maxIOB = self.settingsManager.preferences.maxIOB
+            self.state.maxCOB = self.settingsManager.preferences.maxCOB
+
+            self.state.eventualBG = self.eventualBGString()
+
+            self.renderContacts()
+        }
+    }
+
+    private func renderContacts() {
+        contacts.forEach { renderContact($0) }
+        workItem = DispatchWorkItem(block: {
+            print("in updateContact, no updates received for more than 5 minutes")
+            self.renderContacts()
+        })
+        DispatchQueue.main.asyncAfter(deadline: .now() + 5 * 60 + 15, execute: workItem!)
+    }
+
+    private func renderContact(_ entry: ContactTrickEntry) {
```

Comment:
this larger functions a straight copy-pastes from the `WatchManager`, it probably deserved a refactoring but I don't feel qualified for that :)